File: backend/internal/models/message.go
package models

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

type MessageStatus string

const (
	MessageStatusSent      MessageStatus = "sent"
	MessageStatusDelivered MessageStatus = "delivered"
	MessageStatusRead      MessageStatus = "read"
)

type Message struct {
	ID            string        `gorm:"primaryKey" json:"id"`
	SenderID      string        `gorm:"not null;index" json:"sender_id"`
	RecipientID   *string       `gorm:"index" json:"recipient_id,omitempty"` // For DMs
	GroupID       *string       `gorm:"index" json:"group_id,omitempty"`     // For group messages
	ReplyToID     *string       `gorm:"index" json:"reply_to_id,omitempty"`  // For replies
	Content       string        `json:"content,omitempty"`
	MediaID       *string       `json:"media_id,omitempty"`
	ForwardedFrom *string       `json:"forwarded_from,omitempty"`            // Original sender name if forwarded
	Latitude      *float64      `json:"latitude,omitempty"`                  // For location sharing
	Longitude     *float64      `json:"longitude,omitempty"`                 // For location sharing
	LocationName  *string       `json:"location_name,omitempty"`             // Optional place name
	ScheduledAt   *time.Time    `gorm:"index" json:"scheduled_at,omitempty"` // For scheduled messages
	Status        MessageStatus `gorm:"default:sent" json:"status"`
	EditedAt      *time.Time    `json:"edited_at,omitempty"`                 // When message was edited
	DeletedAt     *time.Time    `json:"deleted_at,omitempty"`                // Soft delete for "delete for everyone"
	ExpiresAt     *time.Time    `gorm:"index" json:"expires_at,omitempty"`   // For disappearing messages
	CreatedAt     time.Time     `json:"created_at"`
	UpdatedAt     time.Time     `json:"updated_at"`

	Sender    User     `gorm:"foreignKey:SenderID" json:"-"`
	Recipient *User    `gorm:"foreignKey:RecipientID" json:"-"`
	Group     *Group   `gorm:"foreignKey:GroupID" json:"-"`
	Media     *Media   `gorm:"foreignKey:MediaID" json:"media,omitempty"`
	ReplyTo   *Message `gorm:"foreignKey:ReplyToID" json:"reply_to,omitempty"`
}

// MessageDeletion tracks "delete for me" operations
type MessageDeletion struct {
	ID        string    `gorm:"primaryKey" json:"id"`
	MessageID string    `gorm:"not null;index;uniqueIndex:idx_msg_user" json:"message_id"`
	UserID    string    `gorm:"not null;index;uniqueIndex:idx_msg_user" json:"user_id"`
	DeletedAt time.Time `json:"deleted_at"`

	Message Message `gorm:"foreignKey:MessageID" json:"-"`
	User    User    `gorm:"foreignKey:UserID" json:"-"`
}

func (md *MessageDeletion) BeforeCreate(tx *gorm.DB) error {
	if md.ID == "" {
		md.ID = uuid.New().String()
	}
	return nil
}

func (m *Message) IsGroupMessage() bool {
	return m.GroupID != nil && *m.GroupID != ""
}

func (m *Message) IsDeleted() bool {
	return m.DeletedAt != nil
}

func (m *Message) IsEdited() bool {
	return m.EditedAt != nil
}

func (m *Message) IsExpired() bool {
	if m.ExpiresAt == nil {
		return false
	}
	return m.ExpiresAt.Before(time.Now())
}

func (m *Message) IsDisappearing() bool {
	return m.ExpiresAt != nil
}

func (m *Message) IsLocation() bool {
	return m.Latitude != nil && m.Longitude != nil
}

func (m *Message) IsScheduled() bool {
	return m.ScheduledAt != nil && m.ScheduledAt.After(time.Now())
}

func (m *Message) BeforeCreate(tx *gorm.DB) error {
	if m.ID == "" {
		m.ID = uuid.New().String()
	}
	if m.Status == "" {
		m.Status = MessageStatusSent
	}
	return nil
}

File: app/lib/models/message.dart
import 'reaction.dart';

enum MessageStatus { sent, delivered, read }

enum MediaType { none, image, video, audio, document }

class ReplyPreview {
  final String id;
  final String senderId;
  final String? content;

  ReplyPreview({
    required this.id,
    required this.senderId,
    this.content,
  });

  factory ReplyPreview.fromJson(Map<String, dynamic> json) {
    return ReplyPreview(
      id: json['id'] as String,
      senderId: json['sender_id'] as String,
      content: json['content'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'sender_id': senderId,
      if (content != null) 'content': content,
    };
  }
}

class Message {
  final String id;
  final String senderId;
  final String? recipientId; // For DMs
  final String? groupId;     // For group messages
  final String? replyToId;   // ID of message being replied to
  final ReplyPreview? replyTo; // Preview of replied message
  final String? content;
  final String? mediaId;
  final String? mediaUrl;
  final String? mediaContentType;
  final MediaType mediaType;
  final String? thumbnailUrl;
  final int? mediaDuration;      // Duration in seconds for audio/video
  final int? mediaWidth;
  final int? mediaHeight;
  final int? pageCount;          // For documents
  final MessageStatus status;
  final DateTime? editedAt;  // When message was edited
  final bool isDeleted;      // Whether message is deleted for everyone
  final List<ReactionInfo> reactions; // Reactions on this message
  final String? forwardedFrom; // Original sender name if forwarded
  final DateTime? expiresAt;   // For disappearing messages
  final DateTime createdAt;

  Message({
    required this.id,
    required this.senderId,
    this.recipientId,
    this.groupId,
    this.replyToId,
    this.replyTo,
    this.content,
    this.mediaId,
    this.mediaUrl,
    this.mediaContentType,
    this.mediaType = MediaType.none,
    this.thumbnailUrl,
    this.mediaDuration,
    this.mediaWidth,
    this.mediaHeight,
    this.pageCount,
    this.status = MessageStatus.sent,
    this.editedAt,
    this.isDeleted = false,
    this.reactions = const [],
    this.forwardedFrom,
    this.expiresAt,
    required this.createdAt,
  });

  bool get isEdited => editedAt != null;
  bool get isDisappearing => expiresAt != null;
  bool get isExpired => expiresAt != null && expiresAt!.isBefore(DateTime.now());
  bool get hasMedia => mediaId != null;
  bool get isImageMessage => mediaType == MediaType.image;
  bool get isVideoMessage => mediaType == MediaType.video;
  bool get isAudioMessage => mediaType == MediaType.audio;
  bool get isDocumentMessage => mediaType == MediaType.document;
  bool get hasReactions => reactions.isNotEmpty;
  bool get isForwarded => forwardedFrom != null;
  String get displayContent => isDeleted ? '[Message deleted]' : (content ?? '');

  factory Message.fromJson(Map<String, dynamic> json) {
    final media = json['media'] as Map<String, dynamic>?;
    return Message(
      id: json['id'] as String,
      senderId: json['sender_id'] ?? json['from'] as String,
      recipientId: json['recipient_id'] as String? ?? json['to'] as String?,
      groupId: json['group_id'] as String?,
      replyToId: json['reply_to_id'] as String?,
      replyTo: json['reply_to'] != null
          ? ReplyPreview.fromJson(json['reply_to'] as Map<String, dynamic>)
          : null,
      content: json['content'] as String?,
      mediaId: json['media_id'] as String?,
      mediaUrl: media?['url'] as String?,
      mediaContentType: media?['content_type'] as String?,
      mediaType: _parseMediaType(media?['media_type'] as String?),
      thumbnailUrl: media?['thumbnail_url'] as String?,
      mediaDuration: media?['duration'] as int?,
      mediaWidth: media?['width'] as int?,
      mediaHeight: media?['height'] as int?,
      pageCount: media?['page_count'] as int?,
      status: parseStatus(json['status'] as String?),
      editedAt: json['edited_at'] != null
          ? DateTime.parse(json['edited_at'] as String)
          : null,
      isDeleted: json['deleted_at'] != null,
      reactions: (json['reactions'] as List<dynamic>?)
              ?.map((r) => ReactionInfo.fromJson(r as Map<String, dynamic>))
              .toList() ??
          [],
      forwardedFrom: json['forwarded_from'] as String?,
      expiresAt: json['expires_at'] != null
          ? DateTime.parse(json['expires_at'] as String)
          : null,
      createdAt: json['created_at'] != null
          ? DateTime.parse(json['created_at'] as String)
          : DateTime.now(),
    );
  }

  static MessageStatus parseStatus(String? status) {
    switch (status) {
      case 'delivered':
        return MessageStatus.delivered;
      case 'read':
        return MessageStatus.read;
      default:
        return MessageStatus.sent;
    }
  }

  static MediaType _parseMediaType(String? type) {
    switch (type) {
      case 'image':
        return MediaType.image;
      case 'video':
        return MediaType.video;
      case 'audio':
        return MediaType.audio;
      case 'document':
        return MediaType.document;
      default:
        return MediaType.none;
    }
  }

  bool get isGroupMessage => groupId != null && groupId!.isNotEmpty;
  bool get isReply => replyToId != null || replyTo != null;

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'sender_id': senderId,
      if (recipientId != null) 'recipient_id': recipientId,
      if (groupId != null) 'group_id': groupId,
      if (replyToId != null) 'reply_to_id': replyToId,
      if (replyTo != null) 'reply_to': replyTo!.toJson(),
      'content': content,
      'media_id': mediaId,
      if (forwardedFrom != null) 'forwarded_from': forwardedFrom,
      'status': status.name,
      'created_at': createdAt.toIso8601String(),
    };
  }

  Message copyWith({
    String? id,
    String? senderId,
    String? recipientId,
    String? groupId,
    String? replyToId,
    ReplyPreview? replyTo,
    String? content,
    String? mediaId,
    String? mediaUrl,
    String? mediaContentType,
    MediaType? mediaType,
    String? thumbnailUrl,
    int? mediaDuration,
    int? mediaWidth,
    int? mediaHeight,
    int? pageCount,
    MessageStatus? status,
    DateTime? editedAt,
    bool? isDeleted,
    List<ReactionInfo>? reactions,
    String? forwardedFrom,
    DateTime? expiresAt,
    DateTime? createdAt,
  }) {
    return Message(
      id: id ?? this.id,
      senderId: senderId ?? this.senderId,
      recipientId: recipientId ?? this.recipientId,
      groupId: groupId ?? this.groupId,
      replyToId: replyToId ?? this.replyToId,
      replyTo: replyTo ?? this.replyTo,
      content: content ?? this.content,
      mediaId: mediaId ?? this.mediaId,
      mediaUrl: mediaUrl ?? this.mediaUrl,
      mediaContentType: mediaContentType ?? this.mediaContentType,
      mediaType: mediaType ?? this.mediaType,
      thumbnailUrl: thumbnailUrl ?? this.thumbnailUrl,
      mediaDuration: mediaDuration ?? this.mediaDuration,
      mediaWidth: mediaWidth ?? this.mediaWidth,
      mediaHeight: mediaHeight ?? this.mediaHeight,
      pageCount: pageCount ?? this.pageCount,
      status: status ?? this.status,
      editedAt: editedAt ?? this.editedAt,
      isDeleted: isDeleted ?? this.isDeleted,
      reactions: reactions ?? this.reactions,
      forwardedFrom: forwardedFrom ?? this.forwardedFrom,
      expiresAt: expiresAt ?? this.expiresAt,
      createdAt: createdAt ?? this.createdAt,
    );
  }
}
